core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableHashJoin.java

```diff
@@ -100,6 +106,45 @@ public static EnumerableHashJoin create(
         condition, variablesSet, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    RelCollation collation = required.getCollation();
+    if (collation == null
+        || collation == RelCollations.EMPTY
+        || joinType == JoinRelType.FULL
+        || joinType == JoinRelType.RIGHT) {
+      return null;
+    }
+
+    List<Integer> requiredKeys = RelCollations.ordinals(collation);
+    ImmutableBitSet requiredKeySet = ImmutableBitSet.of(requiredKeys);
+    ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
+
+    // EnumerableHashJoin traits passdown shall only pass through collation to left input.
+    // It is because for EnumerableHashJoin always builds hash table on right input,
+    // and only non-hashed side can preserve ordering.
+
+    // Only consider exact key match for now
+    if (requiredKeySet.equals(leftKeySet)) {
+      RelTraitSet passthroughTraitSet = traitSet.replace(collation);
+      return Pair.of(passthroughTraitSet,
+          ImmutableList.of(
+              passthroughTraitSet,
+              passthroughTraitSet.replace(RelCollations.EMPTY)));
+    }
+
+    // TODO: consider super keyset and sub keyset
+    return null;
+  }
+
+  @Override public DeriveMode getDeriveMode() {
+    if (joinType == JoinRelType.FULL || joinType == JoinRelType.RIGHT) {
+      return DeriveMode.PROHIBITED;
+    }
+
+    return DeriveMode.BOTH;
```

Comment:
Unless you want to support distribution, otherwise, `LEFT_FIRST` should be the only choice.

Reply:
Sure. let me make a change.  I haven't fully understood the differnet of `LEFT_FIRST`, `RIGHT_FIRST` and `BOTH`.

Reply:
I need to consider refining the java doc to make it easier to understand.